### DISCUSSION TITLE
Periodic recurrence of "Author URL is not accessible" in Site Health

### DISCUSSION THREAD

**Discussion by craigconstantine** [Automattic/wordpress-activitypub#226] [Q&A] [answered]
The Site Health check for accessibility of author URL periodically comes up. Starting this discussion both to seek help, and to start making a trail of crumbs as I discover them.

WP 6.1.1, ActivityPub 0.14.3

I know this happened once before—perhaps more than a week ago.

<img width="834" alt="Screenshot 2022-12-22 at 11 47 58" src="https://user-images.githubusercontent.com/2278992/209185175-362f8e5d-d2f1-45ec-bb78-d75a3c82fc34.png">

Checking with curl, the desired content-type is truly missing. https://moversmindset.com/author/craig/

```
$ curl -i https://constantine.name/author/craig/
HTTP/2 200
cache-control: no-cache
wpo-cache-status: cached
last-modified: Sat, 17 Dec 2022 12:34:24 GMT
vary: Accept-Encoding
content-type: text/html; charset=UTF-8
date: Thu, 22 Dec 2022 16:48:28 GMT
server: Apache

{"@context":["https:\/\/www.w3.org\/ns\/activitystreams","https:\/\/w3id.org\/security\/v1",[[SNIP]]
```

## This fixes it (temporarily?)

I recall that the last time I saw this notice in Site Health, I cleared the pages cache (via the WP-Optimize plugin.)

So I cleared the cache and the Site Health test is now happy. But curl'ing the author URL still does not show the Site Health's desired headers…

```
$ curl -i https://constantine.name/author/craig/
HTTP/2 200
cache-control: no-cache
wpo-cache-status: cached
last-modified: Sat, 17 Dec 2022 12:34:24 GMT
vary: Accept-Encoding
content-type: text/html; charset=UTF-8
date: Thu, 22 Dec 2022 16:51:54 GMT
server: Apache

{"@context":["https:\/\/www.w3.org\/ns\/activitystreams","https:\/\/w3id.org\/security\/v1",[[SNIP]]
```

## First question:

Should I be including an additional header with my curl request?

My curl after clearing the cache still does not show the `application/activity+json` header. So I think my hand-testing with curl isn't actually testing the right thing.

## Second question:

Should I be excluding the `/author/*` path from caching?

**Comment by pfefferle** [comment, marked as answer]
You missed the Accept header in your curl:

```
curl "https://moversmindset.com/author/craig/" \
     -H 'Accept: application/activity+json'
```

But it really looks like a caching issue. The first version, either HTML or JSON, that was requested seems to be cached and returned. Your request (without the Accept header) should always return the HTML version and with the except header it should always return the JSON. Sadly most caching plugins does not support [Content Negotiation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation) :(

So yes, you have to exclude the author pages from the caching.

**Reply by craigconstantine** [reply]
Yeup, disabling caching resolves it. It's nice that the Site Health test caught this!

I told WP-Optimize not to cache the `/author/*` path and now I get the `application/activity+json` in the response `content-type` header:

```
$ curl -i "https://moversmindset.com/author/craig/" -H 'Accept: application/activity+json'
HTTP/2 200
wpo-cache-status: not cached
wpo-cache-message: In the settings, caching is disabled for matches for the current URL
link: <https://moversmindset.com/wp-json/>; rel="https://api.w.org/", <https://moversmindset.com/wp-json/wp/v2/users/1>; rel="alternate"; type="application/json"
vary: User-Agent
content-type: application/activity+json
date: Thu, 22 Dec 2022 20:01:19 GMT
server: Apache

{"@context":["https:\/\/www.w3.org\/ns\/activitystreams","https:\/\/w3id.org\/security\/v1",{"manuallyApprovesFollowers":"as:manuallyApprovesFollowers","PropertyValue":"schema:PropertyValue","schema":"http:\/\/schema.org#","pt":"https:\/\/joinpeertube.org\/ns#","toot":"http:\/\/joinmastodon.org\/ns#","value":"schema:value","Hashtag":"as:Hashtag","featured":{"@id":"toot:featured","@type":"@id"},"featuredTags":{"@id":"toot:featuredTags","@type":"@id"}}],"id":"https:\/\/moversmindset.com\/author\/craig\/","type":"Person","name":"Craig Constantine","summary":"In the podcast, I talk with movement enthusiasts to learn who they are, what they do, and why they do it. I\u2019m interested in the nature and philosophy of movement and in exploring themes like independence, self-direction, and human excellence. My interests color each conversation and provide some structure to Movers Mindset. But since I like to take the scenic route, every conversation ends up going somewhere unique.","preferredUsername":"craig","url":"https:\/\/moversmindset.com\/author\/craig\/","icon":{"type":"Image","url":"https:\/\/secure.gravatar.com\/avatar\/e48a7f64b0279de5efc413a19469bed4?s=120\u0026d=mm\u0026r=g"},"inbox":"https:\/\/moversmindset.com\/wp-json\/activitypub\/1.0\/users\/1\/inbox","outbox":"https:\/\/moversmindset.com\/wp-json\/activitypub\/1.0\/users\/1\/outbox","followers":"https:\/\/moversmindset.com\/wp-json\/activitypub\/1.0\/users\/1\/followers","following":"https:\/\/moversmindset.com\/wp-json\/activitypub\/1.0\/users\/1\/following","manuallyApprovesFollowers":false,"publicKey":{"id":"https:\/\/moversmindset.com\/author\/craig\/#main-key","owner":"https:\/\/moversmindset.com\/author\/craig\/","publicKeyPem":"-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAn11UZQRqL1Zsa0Ly2y33\nZG9S5OsEa6HTCXmmF7aovz4cP+QAQr0hmv1fQYCJ1TdYXYe5C5gypKKW5I4SUc2k\njAgnfvThYd5IC1L+dXxKPPMREXX7FnmxpxAS16eJP\/k\/tRIHxbHf4ysfBvm29HBv\nKw+l+UWGpFmRaPlqytiKxlLR18bIKREP\/1D2fEtKF3qY335WVuQVJ\/APLDJDp6DL\nmQXNDnXjBm56RwKqFxgyiSN14OY7i20pahCeEZyK1fXntHgTS2knBiF6aT5qtQUX\niS8IJmaOk7LHkzZ80w\/1ycZuO36Xb29HJnl2R9UPlunxhIZp1oBlXT3VYKhhDzC0\npwIDAQAB\n-----END PUBLIC KEY-----"},"tag":[],"attachment":[{"type":"PropertyValue","name":"Blog","value":"\u003Ca rel=\u0022me\u0022 title=\u0022https:\/\/moversmindset.com\/\u0022 target=\u0022_blank\u0022 href=\u0022https:\/\/moversmindset.com\/\u0022\u003Emoversmindset.com\u003C\/a\u003E"},{"type":"PropertyValue","name":"Profile","value":"\u003Ca rel=\u0022me\u0022 title=\u0022https:\/\/moversmindset.com\/author\/craig\/\u0022 target=\u0022_blank\u0022 href=\u0022https:\/\/moversmindset.com\/author\/craig\/\u0022\u003Emoversmindset.com\u003C\/a\u003E"}]}
```

**Reply by pfefferle** [reply]
Maybe I should add a complete curl example to site-health, to be able to better debug it?

**Reply by craigconstantine** [reply]
That would have saved me one level of digging. But I wouldn't have guessed caching without that being pointed out to me. So may as well mention caching too.

Bigger idea: What about expanding the documentation (one of those GitHub based web sites that I've seen, but don't understand actaully) Then start a troubleshooting section. Then the Site Health tests can link to a troubleshooting page. (I know that's not trivial. :)
